File: C/Data Structures/Trees/05_Iterative_preorder.c
// IMPLEMENTATION OF ITERATIVE PRE-ORDER TRAVERSAL
// PRE-ORDER CRITERIA: ROOT LEFT RIGHT

#include <stdio.h>
#include <stdlib.h>
#define MAX 20

// create tree node
struct Node {
  int data;
  struct Node *left;
  struct Node *right;
};

// create stack
struct Stack {
  struct Node *stack[MAX];
  int top;
};

// create tree nodes
struct Node *initNode(int item) {
  struct Node *newnode;
  newnode = (struct Node *)malloc(sizeof(struct Node));
  newnode->data = item;
  newnode->left = NULL;
  newnode->right = NULL;
  return newnode;
}

// initialize stack
void initStack(struct Stack *s) { s->top = -1; }

// push tree node into stack
void push(struct Stack *s, struct Node *node) {
  if (s->top > MAX - 1) {
    printf("OVERFLOW!\n");
    exit(1);
  }
  s->stack[++s->top] = node;
}

// pop tree node from the stack
struct Node *pop(struct Stack *s) {
  if (s->top < 0) {
    printf("UNDERFLOW!\n");
    exit(1);
  }
  return s->stack[s->top--];
}

// if stack is empty
int is_empty(struct Stack *s) { return s->top == -1; }

// iterative pre order
void iterative_preorder(struct Node *rootnode) {
  struct Stack s;
  struct Node *node;
  initStack(&s);

  if (!rootnode)
    return;
  push(&s, rootnode);

  printf("Iterative Preorder: [ ");
  while (!is_empty(&s)) {
    node = pop(&s);
    printf("%d ", node->data);
    if (node->right)
      push(&s, node->right);
    if (node->left)
      push(&s, node->left);
  }
  printf("]\n");
}

// main
int main() {
  struct Node *root;
  root = initNode(10);
  root->left = initNode(20);
  root->right = initNode(30);
  root->left->left = initNode(40);
  root->left->right = initNode(50);
  root->right->left = initNode(60);
  root->right->right = initNode(70);
  iterative_preorder(root);
  return 0;
}

File: Python/Data Structures/Trees/05_Iterative_preorder.py
# IMPLEMENTATION OF ITERATIVE PREORDER TRAVERSAL

# create tree node
class Node:
    def __init__(self, item) -> None:
        self.left = None
        self.right = None
        self.data = item


# Implement stack
class Stack:
    def __init__(self) -> None:
        self.stack = []
        self.top = -1

    # push tree node into stack
    def push(self, node):
        self.top += 1
        self.stack.append(node)

    # pop tree node from the stack
    def pop(self):
        temp = self.stack.pop()
        self.top -= 1
        return temp

    def is_empty(self):
        return self.top < 0


def iterative_preorder(root):
    """
    Since stack follows FIFO we need to push right node first
    and then left node.
    """
    s = Stack()

    if not root:
        return
    s.push(root)

    print("Pre-order: ", end="")
    while not s.is_empty():
        node = s.pop()
        print(node.data, end=" ")
        if node.right:
            s.push(node.right)
        if node.left:
            s.push(node.left)
    print("")


# main
def main():
    root = Node(10)
    root.left = Node(20)
    root.right = Node(30)
    root.left.left = Node(40)
    root.left.right = Node(50)
    root.right.left = Node(60)
    root.right.right = Node(70)
    iterative_preorder(root)


if __name__ == "__main__":
    main()
